drivers/epd_bw_spi/epd_bw_spi.c

```diff
@@ -0,0 +1,321 @@
+            return EPD_BW_SPI_BUSY_FAIL;
+        }
+    }
+
+    if (dev->params.dc_pin == GPIO_UNDEF ||
+        gpio_init(dev->params.dc_pin, GPIO_OUT) != 0) {
+        DEBUG("[epd_bw_spi] init: error initializing the DC pin\n");
+        return EPD_BW_SPI_DC_FAIL;
+    }
+    gpio_set(dev->params.dc_pin);
+
+    int res = spi_init_cs(dev->params.spi, dev->params.cs_pin);
+    if (res != SPI_OK) {
+        DEBUG("[epd_bw_spi] init: error initializing the CS pin [%i]\n", res);
+        return res;
+    }
+
+    return 0;
+}
+
+void epd_bw_spi_display_init(epd_bw_spi_t *dev)
+{
+    le_uint16_t y_data[2] = { 0 };
+    uint8_t y_size;
+
+    if (dev->controller.size_y <= 255) {
+        y_data[0].u8[0] = dev->size_y - 1;
+        y_size = 2;
+    }
+    else {
+        y_data[0] =
+            byteorder_btols(byteorder_htons((dev->size_y - 1) & 0x01FF));
+        y_size = 3;
+    }
+    epd_bw_spi_write_cmd(&dev->params, EPD_BW_SPI_CMD_DRIVER_OUTPUT_CONTROL,
+                         (uint8_t *)y_data, y_size);
+
+    uint8_t data[] = {
+        0xD7,   /* Phase 1: 30 ms phase, sel 3, 6.58 us off */
+        0xD6,   /* Phase 2: 30 ms phase, sel 3, 3.34 us off */
+        0x9D,   /* Phase 3: 10 ms phase, sel 4, 1.54 us off */
+    };
+    epd_bw_spi_write_cmd(&dev->params,
+                         EPD_BW_SPI_CMD_BOOSTER_SOFT_START_CONTROL, data, 3);
+
+    data[0] = dev->controller.vcom;
+    epd_bw_spi_write_cmd(&dev->params, EPD_BW_SPI_CMD_WRITE_VCOM_REGISTER, data,
+                         1);
+
+    data[0] = 0x1A; /* 4 dummy line per gate */
+    epd_bw_spi_write_cmd(&dev->params, EPD_BW_SPI_CMD_SET_DUMMY_LINE_PERIOD,
+                         data, 1);
+
+    data[0] = 0x08; /* 2 µs per line */
+    epd_bw_spi_write_cmd(&dev->params, EPD_BW_SPI_CMD_SET_GATE_LINE_WIDTH, data,
+                         1);
+
+    data[0] = (uint8_t)dev->entry_mode;
+    epd_bw_spi_write_cmd(&dev->params, EPD_BW_SPI_CMD_DATA_ENTRY_MODE_SETTING,
+                         data, 1);
+}
+
+void epd_bw_spi_activate(epd_bw_spi_t *dev)
+{
+    uint8_t data[] = { 0xC0 }; /* Enable Clock Signal, Enable CP */
+
+    epd_bw_spi_write_cmd(&dev->params, EPD_BW_SPI_CMD_DISPLAY_UPDATE_CONTROL_2,
+                         data, sizeof data);
+    epd_bw_spi_write_cmd(&dev->params, EPD_BW_SPI_CMD_MASTER_ACTIVATION, NULL,
+                         0);
+    epd_bw_spi_wait(&dev->params, EPD_BW_SPI_WAIT_ACTIVATION);
+}
+
+void epd_bw_spi_deactivate(epd_bw_spi_t *dev)
+{
+    uint8_t data[] = { 0x03 }; /* Disable CP, Disable Clock Signal */
+
+    epd_bw_spi_write_cmd(&dev->params, EPD_BW_SPI_CMD_DISPLAY_UPDATE_CONTROL_2,
+                         data, sizeof data);
+    epd_bw_spi_write_cmd(&dev->params, EPD_BW_SPI_CMD_MASTER_ACTIVATION, NULL,
+                         0);
+    epd_bw_spi_wait(&dev->params, EPD_BW_SPI_WAIT_ACTIVATION);
+}
+
+void epd_bw_spi_init_full(epd_bw_spi_t *dev)
+{
+    epd_bw_spi_display_init(dev);
+    epd_bw_spi_set_area(dev, 0, 200, 0, 200);
```

Comment:
Shouldn't it be `dev->size_`? :eyes:

Also should this be called before setting the look-up table? In various drivers for Waveshare e-Paper displays they seem to do it later. The one I'm using to test is a [1.54 in e-Ink](https://www.waveshare.com/wiki/1.54inch_e-Paper_Module) with source code [here](https://github.com/waveshare/e-Paper/) (I haven't made it work with this driver, yet).

Reply:
Yes, it should be.

I don't think the initialization order here matters. The most important thing is that both registers are initialized. I have tested this driver with two displays, one of them being the Waveshare display and the other one being one with a missing busy/reset pin.